<a href="https://colab.research.google.com/github/eliasjrdg/electronic_store_data_analysis/blob/main/electronic_store_data_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Electronic Store Sales Data**
**Ajude a resolver os problemas reais de negócio**

O projeto envolve processos de importação, processamento, limpeza e análise exploratória de dados.

**Questões a serem respondidas:**

*   *Questão 1:* Qual foi o melhor mês para vendas?
*   *Questão 2:* Qual cidade vendeu mais produtos?
*   *Questão 3:* A que horas devemos exibir anúncios para maximizar a probabilidade de compra do cliente?
*   *Questão 4:* Quais produtos são mais vendidos juntos?
*   *Questão 5:* Qual produto vendeu mais?


***Sobre o Dataset***

O dataset contém dados de vendas de uma loja de eltrônicos por 12 meses em 12 arquivos csv diferentes.

Dataset disponível em: https://www.kaggle.com/datasets/saumaydhaundiyal/electronic-store-sales-data



---



## ***Imports***

In [ ]:
pip install plotly

In [ ]:
import os
import glob
import pandas as pd

In [ ]:
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

### ***Loading data***

In [ ]:
files = glob.glob(os.path.join(os.getcwd(),'sales_data', '*.csv'))

In [ ]:
df = pd.DataFrame()

for x in files:
  df_x = pd.read_csv(x)
  df = pd.concat([df, df_x], ignore_index=True)

df

Order ID                     Product Quantity Ordered Price Each  \
0        259358      34in Ultrawide Monitor                1     379.99   
1        259359      27in 4K Gaming Monitor                1     389.99   
2        259360      AAA Batteries (4-pack)                2       2.99   
3        259361            27in FHD Monitor                1     149.99   
4        259362            Wired Headphones                1      11.99   
...         ...                         ...              ...        ...   
186845   319666    Lightning Charging Cable                1      14.95   
186846   319667       AA Batteries (4-pack)                2       3.84   
186847   319668             Vareebadd Phone                1        400   
186848   319669            Wired Headphones                1      11.99   
186849   319670  Bose SoundSport Headphones                1      99.99   

            Order Date                           Purchase Address  
0       10/28/19 10:56            609 Cherry St, Dallas, TX 75001  
1       10/28/19 17:26          225 5th St, Los Angeles, CA 90001  
2       10/24/19 17:20       967 12th St, New York City, NY 10001  
3       10/14/19 22:26  628 Jefferson St, New York City, NY 10001  
4       10/07/19 16:10         534 14th St, Los Angeles, CA 90001  
...                ...                                        ...  
186845  12/11/19 20:58     14 Madison St, San Francisco, CA 94016  
186846  12/01/19 12:01       549 Willow St, Los Angeles, CA 90001  
186847  12/09/19 06:43           273 Wilson St, Seattle, WA 98101  
186848  12/03/19 10:39             778 River St, Dallas, TX 75001  
186849  12/21/19 21:45     747 Chestnut St, Los Angeles, CA 90001  

[186850 rows x 6 columns]

## ***Data Cleansing***

### Verificando e removendo registros vazios

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186850 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
dtypes: object(6)
memory usage: 8.6+ MB


In [ ]:
df.isna().sum()

Order ID            545
Product             545
Quantity Ordered    545
Price Each          545
Order Date          545
Purchase Address    545
dtype: int64

In [ ]:
print(186850 - 186305)

545


In [ ]:
df = df.dropna()
df.isna().sum()

Order ID            0
Product             0
Quantity Ordered    0
Price Each          0
Order Date          0
Purchase Address    0
dtype: int64

### Verificando e removendo duplicatas

In [ ]:
df.duplicated().value_counts()

False    185687
True        618
Name: count, dtype: int64

In [ ]:
df[df.duplicated(keep=False)]

Order ID                 Product  Quantity Ordered  Price Each  \
521     Order ID                 Product  Quantity Ordered  Price Each   
833     Order ID                 Product  Quantity Ordered  Price Each   
978       260286   AA Batteries (4-pack)                 1        3.84   
979       260286   AA Batteries (4-pack)                 1        3.84   
1452      260739  AAA Batteries (4-pack)                 1        2.99   
...          ...                     ...               ...         ...   
185084    317971   AA Batteries (4-pack)                 1        3.84   
185085    317971   AA Batteries (4-pack)                 1        3.84   
185481  Order ID                 Product  Quantity Ordered  Price Each   
185925  Order ID                 Product  Quantity Ordered  Price Each   
185955  Order ID                 Product  Quantity Ordered  Price Each   

            Order Date                          Purchase Address  
521         Order Date                          Purchase Address  
833         Order Date                          Purchase Address  
978     10/25/19 18:18     529 River St, New York City, NY 10001  
979     10/25/19 18:18     529 River St, New York City, NY 10001  
1452    10/09/19 23:06             372 9th St, Seattle, WA 98101  
...                ...                                       ...  
185084  12/17/19 18:39  250 Chestnut St, San Francisco, CA 94016  
185085  12/17/19 18:39  250 Chestnut St, San Francisco, CA 94016  
185481      Order Date                          Purchase Address  
185925      Order Date                          Purchase Address  
185955      Order Date                          Purchase Address  

[883 rows x 6 columns]

* É possível perceber que vários registros possuem o nome das colunas.
Vamos começar excluindo esses registros.

In [ ]:
df = df[df['Order ID'] != 'Order ID']

In [ ]:
df.duplicated().value_counts()

False    185686
True        264
Name: count, dtype: int64

In [ ]:
df[df.duplicated(keep=False)]

Order ID                 Product Quantity Ordered Price Each  \
978      260286   AA Batteries (4-pack)                1       3.84   
979      260286   AA Batteries (4-pack)                1       3.84   
1452     260739  AAA Batteries (4-pack)                1       2.99   
1453     260739  AAA Batteries (4-pack)                1       2.99   
2337     261586  AAA Batteries (4-pack)                1       2.99   
...         ...                     ...              ...        ...   
182973   315955         ThinkPad Laptop                1     999.99   
183199   316173  AAA Batteries (4-pack)                1       2.99   
183200   316173  AAA Batteries (4-pack)                1       2.99   
185084   317971   AA Batteries (4-pack)                1       3.84   
185085   317971   AA Batteries (4-pack)                1       3.84   

            Order Date                          Purchase Address  
978     10/25/19 18:18     529 River St, New York City, NY 10001  
979     10/25/19 18:18     529 River St, New York City, NY 10001  
1452    10/09/19 23:06             372 9th St, Seattle, WA 98101  
1453    10/09/19 23:06             372 9th St, Seattle, WA 98101  
2337    10/11/19 20:03        703 Main St, Los Angeles, CA 90001  
...                ...                                       ...  
182973  12/26/19 17:28        588 Chestnut St, Seattle, WA 98101  
183199  12/22/19 22:44         907 Sunset St, Portland, OR 97035  
183200  12/22/19 22:44         907 Sunset St, Portland, OR 97035  
185084  12/17/19 18:39  250 Chestnut St, San Francisco, CA 94016  
185085  12/17/19 18:39  250 Chestnut St, San Francisco, CA 94016  

[528 rows x 6 columns]

* Agora vamos remover os outros valores duplicados.

In [ ]:
df = df.drop_duplicates()
df.duplicated().value_counts()

False    185686
Name: count, dtype: int64

### Preparando os tipos de dados

In [ ]:
df['Order ID'] = df['Order ID'].astype(int)
df['Quantity Ordered'] = df['Quantity Ordered'].astype(int)

df['Price Each'] = df['Price Each'].astype(float)

df['Order Date'] = pd.to_datetime(df['Order Date'], format='%m/%d/%y %H:%M')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 185686 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Order ID          185686 non-null  int64         
 1   Product           185686 non-null  object        
 2   Quantity Ordered  185686 non-null  int64         
 3   Price Each        185686 non-null  float64       
 4   Order Date        185686 non-null  datetime64[ns]
 5   Purchase Address  185686 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 9.9+ MB


## ***EDA - Exploratory Data Analysis***

#### **Questão 1:** Qual foi o melhor mês para vendas?

In [ ]:
# separando os pedidos e quantiades
month_sales = df[['Order ID', 'Quantity Ordered']].copy()

# separando o valor em $ de cada pedido para comparar com a quantidade
# podem haver meses com menor quantidade de vendas porém com lucros mais altos
month_sales['Revenue'] = df['Quantity Ordered'] * df['Price Each']

# agrupando pedidos por 'order id' e somando as quantidades e valores
month_sales = month_sales.groupby(df['Order Date'].dt.month).agg({'Revenue': 'sum', 'Quantity Ordered': 'sum'}).reset_index()
month_sales

Order Date     Revenue  Quantity Ordered
0            1  1821413.16             10893
1            2  2200078.08             13431
2            3  2804973.35             16979
3            4  3389217.98             20536
4            5  3150616.23             18653
5            6  2576280.15             15234
6            7  2646461.32             16054
7            8  2241083.37             13429
8            9  2094465.69             13091
9           10  3734777.86             22669
10          11  3197875.05             19769
11          12  4608295.70             28074

In [ ]:
fig = make_subplots(specs=[[{"secondary_y": True}]])

# 'Quantity Ordered' bar plot
fig.add_trace(
    go.Bar(x=month_sales['Order Date'], y=month_sales['Quantity Ordered'], name='Quantity Ordered'), secondary_y=False)

# 'Revenue' line plot
fig.add_trace(
    go.Scatter(x=month_sales['Order Date'], y=month_sales['Revenue'], name='Revenue'), secondary_y=True)


# Title
fig.update_layout(
    title=dict(text='Month Sales and Revenue', font=dict(size=50)))

# X axes
fig.update_xaxes(title_text='Month')

# Y axes
fig.update_yaxes(title_text='Quantity Ordered', secondary_y=False)
fig.update_yaxes(title_text='Revenue', secondary_y=True)

fig.show()

**R:** Dezembro foi o melhor mês para vendas (28.074 produtos vendidos e $4,608,295.70 em lucro)

#### **Questão 2:** Qual cidade vendeu mais produtos?

In [ ]:
city_sales = pd.DataFrame()

# usando a função lambda para selecionar apenas a cidade em 'Purchase Address'
city_sales['city'] = df['Purchase Address'].apply(lambda x: x.split(sep=', ')[1]).copy()

city_sales['sales'] = df['Quantity Ordered']

# agrupando por cidade e somando a quantidade vendida
city_sales = city_sales.groupby('city')['sales'].sum()
city_sales

city
Atlanta          16584
Austin           11137
Boston           22494
Dallas           16707
Los Angeles      33247
New York City    27903
Portland         14037
San Francisco    50169
Seattle          16534
Name: sales, dtype: int64

In [ ]:
fig = px.bar(city_sales)

fig.update_layout(
    title=dict(text='City Product Sales', font=dict(size=50), automargin=True),
    xaxis_title="City",
    yaxis_title="Products")

fig.show()

**R:** San Francisco foi a cidade com o maior número de vandas (50.169)

#### **Questão 3:** A que horas devemos exibir anúncios para maximizar a probabilidade de compra do cliente?

In [ ]:
hour_sales = pd.DataFrame()

# selecionando apenas a hora das vendas
hour_sales['hour'] = df['Order Date'].dt.hour

# selecionando a quantidade de vendas
hour_sales['sales'] = df['Quantity Ordered']

# selecionando o valor $ de cada venda
hour_sales['revenue'] = df['Quantity Ordered'] * df['Price Each']

# agrupando por 'hour' e somando a quantidade de vendas e valores $
hour_sales = hour_sales.groupby('hour').agg({'sales': 'sum', 'revenue': 'sum'}).reset_index()
hour_sales

hour  sales     revenue
0      0   4420   713548.66
1      1   2616   460563.90
2      2   1397   234839.45
3      3    927   145742.94
4      4    935   162546.07
5      5   1492   230664.87
6      6   2809   447963.01
7      7   4547   744067.32
8      8   6998  1192309.20
9      9   9808  1638784.04
10    10  12292  1942988.08
11    11  13986  2296619.84
12    12  14188  2314359.85
13    13  13671  2152369.98
14    14  12343  2082513.70
15    15  11375  1940623.20
16    16  11636  1902523.20
17    17  12214  2126553.97
18    18  13785  2218374.01
19    19  14451  2411971.14
20    20  13758  2280784.36
21    21  12228  2040790.48
22    22   9884  1606243.85
23    23   7052  1177792.82

In [ ]:
fig = make_subplots(specs=[[{"secondary_y": True}]])

fig.add_trace(
    go.Bar(x=hour_sales['hour'], y=hour_sales['sales'], name='Sales'), secondary_y=False)

fig.add_trace(
    go.Scatter(x=hour_sales['hour'], y=hour_sales['revenue'], name='Revenue'), secondary_y=True)

# Title
fig.update_layout(
    title=dict(text='Hour Sales and Revenue', font=dict(size=50)))

# X axes
fig.update_xaxes(title_text="Hour")

# Y axes
fig.update_yaxes(title_text="Sales", secondary_y=False)
fig.update_yaxes(title_text="Revenue", secondary_y=True)

fig.show()

**R:** Os horários de 11h às 13h e de 18h às 20h são os melhores horários para exibir anúncios.

#### **Questão 4:** Quais produtos são mais vendidos juntos?

In [ ]:
# selecionando apenas as ordens com mais de um produto vendido
prod_sales = df[df['Order ID'].duplicated(keep = False)].copy()

# agrupando os produtos por 'order id' e concatenando os nomes dos produtos
prod_sales = prod_sales.groupby('Order ID').agg({'Product': lambda x: ' + '.join(x)}).reset_index()

# contando e selecionando os top 10 grupos de produtos mais vendidos
prod_sales = prod_sales['Product'].value_counts().head(10)

prod_sales

Product
iPhone + Lightning Charging Cable                         886
Google Phone + USB-C Charging Cable                       857
iPhone + Wired Headphones                                 361
Vareebadd Phone + USB-C Charging Cable                    312
Google Phone + Wired Headphones                           303
iPhone + Apple Airpods Headphones                         286
Google Phone + Bose SoundSport Headphones                 161
Vareebadd Phone + Wired Headphones                        104
Google Phone + USB-C Charging Cable + Wired Headphones     79
Vareebadd Phone + Bose SoundSport Headphones               60
Name: count, dtype: int64

In [ ]:
fig = px.bar(prod_sales, x='count')

fig.update_layout(
    title=dict(text='Products Sold Together', font=dict(size=50), automargin=True),
    xaxis_title='Sales',
    yaxis_title='',
    yaxis=dict(autorange='reversed')
)

fig.show()

**R:** 'iPhone + Lightning Charging Cable' são os produtos mais vendidos juntos (886 vendas), seguido por 'Google Phone + USB-C Charging Cable' (857 vendas).

#### **Questão 5:** Qual produto vendeu mais?

In [ ]:
# selecionando produtos e quantidades vendidas
product_sales = df[['Product', 'Quantity Ordered']].copy()

# selecioando o valor $ de cada venda
product_sales['revenue'] = df['Quantity Ordered'] * df['Price Each']

# agrupando por produto e somando as quantidades e valores $
product_sales = product_sales.groupby('Product').agg({'Quantity Ordered': 'sum', 'revenue': 'sum'}).reset_index()

product_sales

Product  Quantity Ordered     revenue
0                 20in Monitor              4126   453818.74
1       27in 4K Gaming Monitor              6239  2433147.61
2             27in FHD Monitor              7541  1131074.59
3       34in Ultrawide Monitor              6192  2352898.08
4        AA Batteries (4-pack)             27615   106041.60
5       AAA Batteries (4-pack)             30986    92648.14
6     Apple Airpods Headphones             15637  2345550.00
7   Bose SoundSport Headphones             13430  1342865.70
8                Flatscreen TV              4813  1443900.00
9                 Google Phone              5529  3317400.00
10                    LG Dryer               646   387600.00
11          LG Washing Machine               666   399600.00
12    Lightning Charging Cable             23169   346376.55
13          Macbook Pro Laptop              4725  8032500.00
14             ThinkPad Laptop              4128  4127958.72
15        USB-C Charging Cable             23931   285975.45
16             Vareebadd Phone              2068   827200.00
17            Wired Headphones             20524   246082.76
18                      iPhone              6847  4792900.00

In [ ]:
fig = make_subplots(specs=[[{"secondary_y": True}]])

fig.add_trace(
    go.Bar(x=product_sales['Product'], y=product_sales['Quantity Ordered'], name='Sales'), secondary_y=False)

fig.add_trace(
    go.Scatter(x=product_sales['Product'], y=product_sales['revenue'], name='Revenue'), secondary_y=True)

# Title
fig.update_layout(
    title=dict(text='Products Sales and Revenue', font=dict(size=50)))

# X axes
fig.update_xaxes(title_text="Product")

# Y axes
fig.update_yaxes(title_text="Sales", secondary_y=False)
fig.update_yaxes(title_text="Revenue", secondary_y=True)

fig.show()

**R:** AAA Batteries(4-pack) é o produto mais vendido em quantidade (30.986 unidades vendidas), porém o produto com maior lucro é Macbook Pro Laptop com (4.725 unidades vendidas e $8,032,500.00 em lucro).